### PR TITLE
refactor fp4 masked gemm cute-dsl implementation and add manual cache

### PR DESCRIPTION
**Pull request by yzh119**
## 📌 Description

The kernel interface exposed in #1331 is not friendly for JIT caching and mix the compile time parameters and runtime arguments. This PR refactors these classes following these rules:

1. Compile time parameters should be passed to `__init__` functions.
2. Use `@cute.kernel` to decorate device-side kernels.
3. Use `@cute.jit` to decorate host-side launcher functions (https://docs.nvidia.com/cutlass/media/docs/pythonDSL/cute_dsl_general/dsl_introduction.html), where we can call device-side kernels.
4. `cute.compile` accepts a host-side launcher function decorated by `@cute.jit`.

## 🔍 Related Issues

https://github.com/flashinfer-ai/flashinfer/issues/1519

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

### CODE REVIEW

File: flashinfer/cute_dsl/blockscaled_gemm.py [hunk 1]
```diff
@@ -2548,183 +2564,189 @@ def ceil_div(a, b):
...
+    alpha_dtype: Optional[Type[cutlass.Numeric]],
+    sf_vec_size: int,
+    mma_tiler_mn: Tuple[int, int],
+    cluster_shape_mn: Tuple[int, int],
+) -> Callable:
+    def get_cute_pointers(
+        input_tensors: Optional[List[torch.tensor]],
+    ) -> List[cute.Pointer]:
+        mock_ptr = torch.randn(16, device="cuda:0").data_ptr()
+        if input_tensors is None:
```
Comment (fzyzcjy) [inline flashinfer/cute_dsl/blockscaled_gemm.py]:
```suggestion
        if input_tensors is None:
            mock_ptr = torch.randn(16, device="cuda:0").data_ptr()
```

Comment (yzh119) [reply flashinfer/cute_dsl/blockscaled_gemm.py]: We don't really need to create the `mock_ptr`, will remove that in later commits.

Comment (fzyzcjy) [reply flashinfer/cute_dsl/blockscaled_gemm.py]: oh I see, looks great

**Review by fzyzcjy — APPROVED** [review]
the initial commit already solves my issue of the 1s host overhead for already seen shapes
